### ISSUE TITLE
1.89 fails in 17 tests on Debian 

### ISSUE THREAD

**Issue by Konfekt** [wofr06/lesspipe.sh#61]
As discussed [here](https://github.com/wofr06/lesspipe/commit/42176ff4e369d4dd78dec638e46dca602857d36d#commitcomment-57567172)

```
less -f testok/a\?b.gz                                   NOT ok
less -f testok/a\[b.gz                                   NOT ok
less -f testok/a\]b.gz                                   NOT ok
```

and

```
less    testok/a\ b.tgz:testok/a\ text.gz                NOT ok
```

but

```
less -f testok/perlstorable.gz   
```

Perhaps it is not a problem of the file or mime type not having been properly detected? @otandka, could you print the output of less on these files (and perhaps that of `lesspipe.sh`)?

**Comment by otandka** [comment]
I include the output of test.pl -d, which should provide the output of all the less commands.  The pipe.[12345] are the output of lesspipe.sh on these files with bash -x enabled.  I think this will make it easier to see where things are going wrong.

```
ota@alfriston:tmp/lesspipe% git log --oneline -n 1
73848a3 (HEAD -> lesspipe, origin/lesspipe, origin/HEAD) use file --mime instead of file -i for compatibility with Mac
```
```
ota@alfriston:tmp/lesspipe% git diff
diff --git a/lesspipe.sh b/lesspipe.sh
index 364390c..6529822 100755
--- a/lesspipe.sh
+++ b/lesspipe.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/bash -x
 # lesspipe.sh, a preprocessor for less (version 1.87)
 #===============================================================================
 ### THIS FILE IS GENERATED FROM lesspipe.sh.in, PLEASE GET THE ZIP FILE
```

```
ota@alfriston:tmp/lesspipe% LESSOPEN="|/tmp/lesspipe/lesspipe.sh %s" ./lesspipe.sh testok/a\[b.gz >& /tmp/pipe.2 ||  echo fail
fail
ota@alfriston:tmp/lesspipe% LESSOPEN="|/tmp/lesspipe/lesspipe.sh %s" ./lesspipe.sh testok/a\]b.gz >& /tmp/pipe.3 ||  echo fail
fail
ota@alfriston:tmp/lesspipe% LESSOPEN="|/tmp/lesspipe/lesspipe.sh %s" ./lesspipe.sh testok/a\?b.gz >& /tmp/pipe.1 ||  echo fail
fail
ota@alfriston:tmp/lesspipe% LESSOPEN="|/tmp/lesspipe/lesspipe.sh %s" ./lesspipe.sh testok/a\ b.tgz:testok/a\ text.gz >& /tmp/pipe.4 ||  echo fail
fail
ota@alfriston:tmp/lesspipe% LESSOPEN="|/tmp/lesspipe/lesspipe.sh %s" ./lesspipe.sh testok/perlstorable.gz >& /tmp/pipe.5 ||  echo fail
ota@alfriston:tmp/lesspipe% git checkout lesspipe.sh

ota@alfriston:tmp/lesspipe% git status --short --branch --untracked-files=no
## lesspipe...origin/lesspipe
ota@alfriston:tmp/lesspipe% LESSOPEN="|/tmp/lesspipe/lesspipe.sh %s" ./test.pl -d >& /tmp/test.pl-d.73848a3.out 
```

HtH

[lesspipe.73848a3.zip](https://github.com/wofr06/lesspipe/files/7298768/lesspipe.73848a3.zip)

**Comment by Konfekt** [comment]
Reading the outputs of the zip file, say `code2color 1841 '-itestok/a text.gz' -` in `pipe4`, [running `code2color`](https://github.com/wofr06/lesspipe/blob/73848a3e4bd9e7ad69a61249e9cd33773fb652c0/lesspipe.sh.in#L690) always returns an error code. What is the output of, say, `code2color '-itestok/a text.gz'` ? (Similar for the other pipes.)

**Comment by Konfekt** [comment]
One obvious reason could be that `testok/a text.gz` exists, but `'-itestok/a text.gz'` not.

However, appending `-i` happens [on purpose](https://github.com/wofr06/lesspipe/blob/73848a3e4bd9e7ad69a61249e9cd33773fb652c0/lesspipe.sh.in#L240). If @wofr06 could explain?

**Comment by otandka** [comment]
The syntax "testok/a\ b.tgz:testok/a\ text.gz" means that a\ text.gz is to be extracted from the (compressed) a\ b.taz file.  It looks to me like it should be extracting the tar file to stdout so it doesn't actually get written to a file, but I'm not sure about that.  I don't know why it is calling code2color on a gziped file.

The number there is the pid of the lesspipe.sh process, which code2color looks up in the process table.

From looking at the code2color code, we see that -i means read from stdin, but assume the file's name was whatever follows the -i.  This is why make test sometimes hangs when it malforms the code2color command or otherwise invokes it wrong.

Similarly, -l<lang> is sometimes appended to the code2color command if lesspipe has inferred the language (in this case it apparently has not).

**Comment by Konfekt** [comment]
The `show` function that [appends `-i`](https://github.com/wofr06/lesspipe/blob/73848a3e4bd9e7ad69a61249e9cd33773fb652c0/lesspipe.sh.in#L240) is somewhat convoluted, and its creator will understand it better than me. Looking at it, I wonder why it used to pass all tests before. @otandka, which was the last commit where all these tests passed? Could you add the output of these passed tests?

**Comment by otandka** [comment]
As I said earlier, the commit tagged 1.88 works fine for me on Debian.  There are only 4 commits between that one and 1.89, which is broken, so I checked them all.  The commit that breaks the tests is 2ad7a65, which comments out the infamous 3 lines.

Here is a zip file with the output of test.pl -d, which includes the result of each run of less that it is checking against.

[lesspipe.test.pl-d.zip](https://github.com/wofr06/lesspipe/files/7316870/lesspipe.test.pl-d.zip)

HtH.

**Comment by wofr06** [comment]
Seems to be fixed in 1.91
